Added throttle for message publishing

pierantoniomerlino opened the pull request:
The idea is to manage the rate and burst of publishing using the Token Bucket algorithm.
The user can configure in the DataService the "Maximum publish rate" and "Maximum number of messages in publish bursts". The first parameter is used to set the maximum messages per seconds that can be published. The second one, instead, sets the number of messages that can be published in a burst at max speed. After the burst, the publishing rate is limited to "Maximum publish rate" value.

If the "Maximum publish rate" is set to 0, no limits are set and the messaged are published at max speed. If "Maximum number of messages in publish bursts" is set to 0, the burst are disabled and the messages are published at "Maximum publish rate". The latter parameter is used only if the "Maximum publish rate" is set to a value different than 0.

Of course, this PR is preliminary and any suggestions are more than welcome!

Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

pierantoniomerlino committed:
Added throttle for message publishing
Modified metatype to include token bucket-related properties.

Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

MMaiero commented on kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml:
```diff
@@ -98,7 +98,25 @@
+
+        <AD id="max.publish.rate"
+            name="Maximum publish rate"
+            type="Double"
```
Why this is a double?

pierantoniomerlino replied:
max.publish.rate is the number of messages that should be send in a second. So, if you want to publish 1 messages every 10 seconds, the value is 0.1. Float is sufficient.
Is there a better way to show it?

MMaiero replied:
Maybe specifying a time unit.

MMaiero commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:
```diff
@@ -104,6 +109,14 @@ protected void activate(ComponentContext componentContext, Map<String, Object> p
 
         this.properties.putAll(properties);
 
+        if (this.properties.get(PUBLISH_RATE) != null && this.properties.get(BURST_LENGTH) != null) {
```
I think that this can be externalized to a separate method.

MMaiero commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:
```diff
@@ -526,6 +536,27 @@ public void run() {
     }
 
+    private void getThrottle() {
+        if (this.properties.get(PUBLISH_RATE) != null && this.properties.get(BURST_LENGTH) != null) {
```
Isn't it more readable if we save `this.properties.get(PUBLISH_RATE)` and `this.properties.get(BURST_LENGTH)` into local variables?

pierantoniomerlino replied:
👍

MMaiero commented:
If we compare our implementation to a token bucket, I think that maybe we should stick with common terms like (https://en.wikipedia.org/wiki/Token_bucket): 

- `Average rate` instead of `Maximum publish rate`

- `Burst size` instead of `Maximum number of messages in publish bursts`

What do you think?

pierantoniomerlino commented:
IMHO, we have to think what is easier to understand from an user point of view.
Is it better to expose the "token bucket" terms or use terms related to the data service?

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:
```diff
@@ -572,14 +598,7 @@ public void run() {
-                        // Notify the listeners
-                        DataServiceImpl.this.dataServiceListeners.onMessagePublished(message.getId(),
-                                message.getTopic());
+                        tryPublishInternal(message);
```
why "try"?

pierantoniomerlino replied:
Because if there aren't tokens in the bucket, the Data service can't publish the message. So, it tries to publish something...

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:
```diff
@@ -593,6 +612,15 @@ public void run() {
 
+    private synchronized void tryPublishInternal(DataMessage message) throws KuraException {
+        if (this.throttle.getToken()) {
+            logger.debug("Throttle: got token - remaining token {}", this.throttle.getRemainingTokens());
```
"remaining tokens"

pierantoniomerlino replied:
👍

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java:
```diff
@@ -0,0 +1,79 @@
+/*******************************************************************************
+ * Copyright (c) 2011, 2017 Eurotech and/or its affiliates
```
only 2017

pierantoniomerlino replied:
👍

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java:
```diff
@@ -0,0 +1,79 @@
+ * Contributors:
+ *     Eurotech
+ *******************************************************************************/
+package org.eclipse.kura.core.data.internal;
```
I believe it should be org.eclipse.kura.core.internal.data. Check the Wires implementation bundles and the Eclipse guidelines for package naming conventions.

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java:
```diff
@@ -0,0 +1,79 @@
+        this.capacity = (capacity == 0) ? 1 : capacity;
+        this.remainingTokens = capacity;
+        this.refillPeriod = refillPeriod;
+        this.lastRefillTime = 0L;
```
Remove "L"

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java:
```diff
@@ -0,0 +1,79 @@
+
+    public boolean getToken() {
+        boolean success;
+        if (this.refillPeriod == 0L) {
```
Remove "L"

cdealti commented on kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml:
```diff
@@ -98,7 +98,36 @@
+            cardinality="0" 
+            required="true"
+            default="messages.second"
+            description="">
```
Add description

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java:
```diff
@@ -0,0 +1,79 @@
+
+public class TokenBucket {
+
+    private long capacity;
```
There will plenty of room just using ``int``.

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java:
```diff
@@ -0,0 +1,79 @@
+public class TokenBucket {
+
+    private long capacity;
+    private long remainingTokens;
```
Same

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java:
```diff
@@ -0,0 +1,79 @@
+    private long lastRefillTime;
+
+    public TokenBucket(long capacity, long refillPeriod) {
+        this.capacity = (capacity == 0) ? 1 : capacity;
```
Unclear

pierantoniomerlino replied:
I'll add a comment about this.

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java:
```diff
@@ -0,0 +1,79 @@
+        return this.capacity;
+    }
+
+    public void setCapacity(long capacity) {
```
Remove setters. Create a new instance instead

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java:
```diff
@@ -0,0 +1,79 @@
+        return this.lastRefillTime;
+    }
+
+    public boolean getToken() {
```
``hasToken`` preferred

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    public boolean getToken() {
+        boolean success;
```
intialize to ``false``

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java:
```diff
@@ -0,0 +1,79 @@
+            if (this.remainingTokens > 0) {
+                this.remainingTokens--;
+                success = true;
+            } else {
```
remove ``else``

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java:
```diff
@@ -0,0 +1,79 @@
+        return success;
+    }
+
+    private void tryRefill() {
```
Just ``refill`` returning the number of remaining tokens.

However, we miss an ``waitToken`` method which blocks until one token is available.
Make sure that the wait can be properly interrupted.

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:
```diff
@@ -156,6 +164,8 @@ public void updated(Map<String, Object> properties) {
         this.properties.clear();
         this.properties.putAll(properties);
 
+        getThrottle();
```
``createThrottle`` or something indicating that an instanciation  will occur

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:
```diff
@@ -526,6 +536,28 @@ public void run() {
     }
 
+    private void getThrottle() {
+        if (this.properties.get(PUBLISH_RATE) != null && this.properties.get(BURST_SIZE) != null) {
```
Hence a new ``TokenBucket`` instance will be created

cdealti commented on kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml:
```diff
@@ -98,7 +98,36 @@
+            min="0"
+            description="The average message publish rate. The unit of measurement is set by the publish rate unit of measurement option. If set to 0 the rate limit is disabled."/>
+        
+        <AD id="average.publish.rate.unit"
```
I believe this is not needed.
I believe that messages/s is the right unit.
Realistically 1 [message/s] is probably what one will set.
Is it too fast for the broker? Setting 0.1 will do the trick

pierantoniomerlino replied:
Yes, but is it enough clear to the user set a rate of 0.1 message/second? If the maximum rate is 1 message/second, so we should change the unit to message/minute and get an integer value.

cdealti replied:
I suggest to add the 0.1 example to the description

pierantoniomerlino replied:
I'll remove the unit option and set the rate as messages/second. I'll also update the description to give an example on how to use it. However, I don't like it.

cdealti commented on kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml:
```diff
@@ -98,7 +98,36 @@
+            required="true"
+            default="0"
+            min="0"
+            description="When the average publish rate is set, this option enables the publishing of messages in bursts at maximum speed. If set to 0 the bursts are disabled."/>
```
Just to double check. The broker will not be happy with bursts. Setting this property to 1 disables bursts entirely?

pierantoniomerlino replied:
The burst size is the capacity of the token bucket. If it is 0, the token bucket constructor force the capacity to 1, so only a token per period can be extracted from the bucket and there will not be bursts.

cdealti replied:
Then set min="1". WYSIWYG

pierantoniomerlino replied:
Ok.

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:
```diff
@@ -593,6 +618,15 @@ public void run() {
     }
 
+    private synchronized void tryPublishInternal(DataMessage message) throws KuraException {
+        if (this.throttle.getToken()) {
```
It should just be ``this.throttle.waitToken()``. Otherwise if this method is called from the publishing task and the DB is full and you don't have tokens, the CPU will spin doing nothing.

With waitToken() at least the thread will block until a token is available.

pierantoniomerlino replied:
Do we need a timeout for this?

cdealti commented:
About the priority of this one. It seems needed to limit the publish rate with Azure IoT but we need to recheck. See #1228.

@markoer  the DataService is a key component that if it fails will cause headaches with customers.
Assuming this will be merged, can you please add some new QA tests for validation covering the corner cases? You should pay attention to cases where the messages DB contains many messages, like after a disconnect and the client reconnects.

cdealti commented:
See #1416.

MMaiero committed:
Modified tocken bucket code.

Added property in data service to explicitly enable/disable the
rate limiting.
Moved to DataServiceOptions.java the properties management.
Thread sleeping is now performed outside the tokenBucket class.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:
```diff
@@ -76,7 +70,7 @@
 
     // A dedicated executor for the publishing task
-    private ScheduledExecutorService publisherExecutor;
+    private ExecutorService publisherExecutor;
```
It looks to me that a plain old Java Thread would me enough

MMaiero replied:
I believe in this context there is no real advantage in one or the other solution.

cdealti commented on kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml:
```diff
@@ -98,7 +98,46 @@
             description="Timeouts the in-flight messages congestion condition. The service will force a disconnect attempting to reconnect (0 to disable)."/>
-            
+
+		<AD id="enable.rate.limit"
```
Fix indentation

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:
```diff
@@ -516,16 +523,27 @@ public void run() {
         return autoConnect;
     }
 
+    private void createThrottle() {
```
Should we add NPE checks for this.throttle everywhere?

MMaiero replied:
Uhm, where do you think it is needed?

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:
```diff
@@ -680,4 +650,99 @@ public CloudConnectionStatusEnum getNotificationStatus() {
+                            if (dataServiceOptions.isRateLimitEnabled() && message.getPriority() >= 5) {
+                                messagePublished = publishMessageTokenBucket(message);
+                            } else {
+                                messagePublished = publishMessageUnbound(message);
```
this method alway returns true so set messagePublished to true explicitly or intialize it to true.

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:
```diff
@@ -680,4 +650,99 @@ public CloudConnectionStatusEnum getNotificationStatus() {
+        private void suspendPublisher(long timeout, TimeUnit timeUnit) throws InterruptedException {
+            try {
+                lock.lock();
+                if (!publisherEnabled) {
```
Why ``publisherEnabled`` must be checked in a lock?
You don't need to throw an Exception, just return.

MMaiero replied:
It's to prevent concurrent changes and reads of the shared variable.

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:
```diff
@@ -680,4 +650,99 @@ public CloudConnectionStatusEnum getNotificationStatus() {
+            }
+        }
+
+        private boolean publishMessageUnbound(DataMessage message) throws KuraException {
```
return void

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceOptions.java:
```diff
@@ -0,0 +1,117 @@
+package org.eclipse.kura.core.data;
```
Copyright header missing

MMaiero replied:
Correct. Missed that.

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/internal/data/TokenBucket.java:
```diff
@@ -0,0 +1,58 @@
+public class TokenBucket {
+
+    private final int capacity;
+    // period in mills between 1 token refill
```
Check spelling or remove comment

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/internal/data/TokenBucket.java:
```diff
@@ -0,0 +1,58 @@
+        return result;
+    }
+
+    private boolean isTokenAvailable() {
```
Move private methods at the bottom

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/internal/data/TokenBucket.java:
```diff
@@ -0,0 +1,58 @@
+            this.remainingTokens = (int) Math.min(this.capacity,
+                    this.remainingTokens + (now - this.lastRefillTime) / this.refillPeriod);
+            this.remainingTokens = Math.max(1, this.remainingTokens);
+            this.lastRefillTime = now;
```
lastRefillTime is not accurate as you lose the elapsed part of the refillPeriod at the time refill() is called.
I believe it should be computed as:

lastRefillTime += ((now - lastRefillTime) / refillPeriod) * refillPeriod

Note that this does not equal ``now`` due to the calculation uses integer arithmetics.

cdealti commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/internal/data/TokenBucket.java:
```diff
@@ -0,0 +1,58 @@
+        }
+    }
+
+    public long getNextRefillEta() {
```
Maybe add a comment that you don't get an updated and meaningful value until getToken() is called. This usage is misleading and error prone however.
The method name should also indicate a "time to next refill" (relative to now) instead of an "ETA of the next refill" (relative to epoch).

pierantoniomerlino commented on kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:
```diff
@@ -680,4 +649,106 @@ public CloudConnectionStatusEnum getNotificationStatus() {
+            boolean executionEnabled;
+            lock.lock();
+            executionEnabled = publisherEnabled;
+            lock.unlock();
```
Is it possible to use an AtomicBoolean instead?

MMaiero replied:
Agreed and done. I've done also some minimal testing and it seems to work.

MMaiero committed:
Modified publisher work.

Now it is a single thread managed by locks, awaits and signals.
Renamed properties in metatype.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>
